### DISCUSSION TITLE
Example identification_of_an_electromechanical_system.ipynb not working

### DISCUSSION THREAD

**Discussion by ghost** [wilsonrljr/sysidentpy#92] [Q&A] [answered]
Hello,

First of all, thanks for the creation of this project, tools for working with NARMAX are not easily available. I am currently evaluating SysIdentPy to be used at work and I am going through some examples and found this example throughs up the following error:

```
ValueError                                Traceback (most recent call last)
Input In [8], in <cell line: 1>()
----> 1 model.fit(X=x_train, y=y_train)
      2 yhat = model.predict(X=x_valid, y=y_valid)
      3 rrse = root_relative_squared_error(y_valid, yhat)

File ~/sysidentpy/sysidentpy/model_structure_selection/forward_regression_orthogonal_least_squares.py:476, in FROLS.fit(self, X, y)
    471 self.regressor_code = self.regressor_space(
    472     self.non_degree, self.xlag, self.ylag, self._n_inputs, self.model_type
    473 )
    475 if self._order_selection is True:
--> 476     self.info_values = self.information_criterion(reg_matrix, y)
    478 if self.n_terms is None and self._order_selection is True:
    479     model_length = np.where(self.info_values == np.amin(self.info_values))

File ~/sysidentpy/sysidentpy/model_structure_selection/forward_regression_orthogonal_least_squares.py:359, in FROLS.information_criterion(self, X_base, y)
    356 n_theta = i + 1
    357 regressor_matrix = self.error_reduction_ratio(X_base, y, n_theta)[2]
--> 359 tmp_theta = getattr(self, self.estimator)(regressor_matrix, y)
    361 tmp_yhat = np.dot(regressor_matrix, tmp_theta)
    362 tmp_residual = y[self.max_lag :] - tmp_yhat

File ~/sysidentpy/sysidentpy/parameter_estimation/estimators.py:255, in Estimators.recursive_least_squares(self, psi, y)
    253 k = np.divide(k_numerator, k_denominator)
    254 self.xi[i, 0] = y[i, 0] - np.dot(psi_tmp.T, theta[:, i - 1])
--> 255 theta[:, i] = list(theta[:, i - 1].reshape(-1, 1) + k.dot(self.xi[i, 0]))
    257 p1 = p.dot(psi[i, :].reshape(-1, 1)).dot(psi[i, :].reshape(-1, 1).T).dot(p)
    258 p2 = (
    259     psi[i, :].reshape(-1, 1).T.dot(p).dot(psi[i, :].reshape(-1, 1))
    260     + self._lam
    261 )

ValueError: setting an array element with a sequence. The requested array would exceed the maximum number of dimension of 1.
```
Could you please give me a clue of how to make it work?

Have python 3.9, SysIdentPy 0.2.0 running on Mac Monterrey

Thanks in advance

**Comment by wilsonrljr** [comment]
Hey ghost! Thanks for reaching out. I'll try to reproduce the error later today. Could you try to run that example using the v0.2.1 so we can check if the error is fixed in the last version?

**Reply by gamcorn** [reply]
Hello Wilson, Thanks for the time you have taken to respond. I wrote the question yesterday but I had to change account so it appears as ghost....

Back to this issue, I have upgraded to v0.2.1, and the problem is still present, then I figured out that this was a Numpy issue I dig further in and install different versions of numpy from 2020 to 2022 (tested around 8 versions) this did not correct the behaviour so I took a look to what is really going on and I fixed the problem by doing the following:

```
line 242 from estimators.py
        for i in range(2, n):
            psi_tmp = psi[i, :].reshape(-1, 1)
            k_numerator = self._lam ** (-1) * p.dot(psi_tmp)
            k_denominator = 1 + self._lam ** (-1) * psi_tmp.T.dot(p).dot(psi_tmp)
            k = np.divide(k_numerator, k_denominator)
            self.xi[i, 0] = y[i, 0] - np.dot(psi_tmp.T, theta[:, i - 1])
            tmp_list = theta[:, i - 1].reshape(-1, 1) + k.dot(self.xi[i, 0])
            theta[:, i] = tmp_list.flatten()
```

I did this on all functions and all the tests on /sysidentpy/parameter_estimation/tests/test_estimators.py are passing now.

I am not 100% certain that what I am doing is the best/right thing to do, if you have better suggestion or if otherwise you want me to submit a PR please let me know.

**Reply by wilsonrljr** [reply]
Hey! No problem regarding your username. I'll investigate this issue this weekend. I tried to replicate it, but I didnt get any error. I'll try again this weekend so I can add any fix in the next update (which is almost done)

**Reply by gamcorn** [reply]
Thanks! if you need any detail from my runs do not hesitate to ask, I will be glad to help !

**Reply by wilsonrljr** [reply]
Hey! I tried different thing here and I got the same error only in a specific case: using numpy 1.24.*. Testing with numpy version 1.23.5 or older worked as expected, however the error appeared using the newest version of numpy. 

You said you tested older versions of numpy and the error persisted. I tested with several different versions as well and the only one with that problem was the 1.24.*. Could you create a new environment with numpy 1.23.5 and test it again, please? In the new environment, check if the version is right in the kernel after installing SysIdentPy just to make sure.

I'll update the requirements in the next version to not use the 1.24* version of numpy. After that I'll upload another version just to fix this issue with the new versions of numpy.

**Reply by gamcorn** [reply, marked as answer]
Yes ! that is correct, version 1.24.x is the one that fails. I did something wrong the last time I checked previous numpy versions and jupyter notebook was not refreshing. So the scenario you describe is correct only version 1.24x has the issue. Would you like me to make a pull request with the changes I made for it to work on above 1.24 ?

**Reply by wilsonrljr** [reply]
Hey @gamcorn , I'll add your fix in the next update. Thank you!
